src/main/scala/is/hail/rvd/OrderedRVD.scala

```diff
@@ -0,0 +1,525 @@
+
+      override def unpersist(): OrderedRVD = {
+        persistedRDD.unpersist()
+        self
+      }
+    }
+  }
+
+  override def cache(): OrderedRVD = persist(StorageLevel.MEMORY_ONLY)
+  
+  override def unpersist(): OrderedRVD = this
+
+  def orderedJoinDistinct(right: OrderedRVD, joinType: String): RDD[JoinedRegionValue] = {
+    val lTyp = typ
+    val rTyp = right.typ
+
+    if (lTyp.kType != rTyp.kType)
+      fatal(
+        s"""Incompatible join keys.  Keys must have same length and types, in order:
+           | Left key type: ${ lTyp.kType.toPrettyString(compact = true) }
+           | Right key type: ${ rTyp.kType.toPrettyString(compact = true) }
+         """.stripMargin)
+
+    joinType match {
+      case "inner" | "left" => new OrderedJoinDistinctRDD2(this, right, joinType)
+      case _ => fatal(s"Unknown join type `$joinType'. Choose from `inner' or `left'.")
+    }
+  }
+
+  def partitionSortedUnion(rdd2: OrderedRVD): OrderedRVD = {
+    assert(typ == rdd2.typ)
+    assert(partitioner == rdd2.partitioner)
+
+    val localTyp = typ
+    OrderedRVD(typ, partitioner,
+      rdd.zipPartitions(rdd2.rdd) { case (it, it2) =>
+        new Iterator[RegionValue] {
+          private val bit = it.buffered
+          private val bit2 = it2.buffered
+
+          def hasNext: Boolean = bit.hasNext || bit2.hasNext
+
+          def next(): RegionValue = {
+            if (!bit.hasNext)
+              bit2.next()
+            else if (!bit2.hasNext)
+              bit.next()
+            else {
+              val c = localTyp.kInRowOrd.compare(bit.head, bit2.head)
+              if (c < 0)
+                bit.next()
+              else
+                bit2.next()
+            }
+          }
+        }
+      })
+  }
+
+  def copy(typ: OrderedRVType = typ,
+    orderedPartitioner: OrderedRVPartitioner = partitioner,
+    rdd: RDD[RegionValue] = rdd): OrderedRVD = {
+    OrderedRVD(typ, orderedPartitioner, rdd)
+  }
+
+  def naiveCoalesce(maxPartitions: Int): OrderedRVD = {
+    val n = partitioner.numPartitions
+    if (maxPartitions >= n)
+      return this
+
+    val newN = maxPartitions
+    val newNParts = Array.tabulate(newN)(i => (n - i + newN - 1) / newN)
+    assert(newNParts.sum == n)
+    assert(newNParts.forall(_ > 0))
+
+    val newPartEnd = newNParts.scanLeft(-1)(_ + _).tail
+    assert(newPartEnd.last == n - 1)
+
+    val newRangeBounds = UnsafeIndexedSeq(
+      TArray(typ.pkType),
+      newPartEnd.init.map(partitioner.rangeBounds))
+
+    OrderedRVD(
+      typ,
+      new OrderedRVPartitioner(newN, typ.partitionKey, typ.kType, newRangeBounds),
+      new BlockedRDD(rdd, newPartEnd))
+  }
+
+  override def coalesce(maxPartitions: Int, shuffle: Boolean): OrderedRVD = {
+    require(maxPartitions > 0, "cannot coalesce to nPartitions <= 0")
+    val n = rdd.partitions.length
+    if (!shuffle && maxPartitions >= n)
+      return this
+    if (shuffle) {
+      val shuffled = rdd.coalesce(maxPartitions, shuffle = true)
```

Comment:
Does this require rewriting like what was needed for persist? Could you please double check I rewrote coalesce properly?

Reply:
Yes, I read over coalesce, it seemed correct to me.